pkg/messenger/messager.go

```diff
@@ -0,0 +1,317 @@
+package messenger
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"gocloud.dev/pubsub"
+	_ "gocloud.dev/pubsub/awssnssqs"
+	_ "gocloud.dev/pubsub/azuresb"
+	_ "gocloud.dev/pubsub/gcppubsub"
+	_ "gocloud.dev/pubsub/kafkapubsub"
+	_ "gocloud.dev/pubsub/natspubsub"
+	_ "gocloud.dev/pubsub/rabbitpubsub"
+)
+
+type Messenger struct {
+	Deployments DeploymentManager
+	Endpoints   EndpointManager
+	Queues      QueueManager
+
+	HTTPC *http.Client
+
+	requests  *pubsub.Subscription
+	responses *pubsub.Topic
+
+	consecutiveErrorsMtx sync.RWMutex
+	consecutiveErrors    int
+}
+
+func NewMessenger(
+	ctx context.Context,
+	requestsURL string,
+	responsesURL string,
+	deployments DeploymentManager,
+	endpoints EndpointManager,
+	queues QueueManager,
+	httpClient *http.Client,
+) (*Messenger, error) {
+	requests, err := pubsub.OpenSubscription(ctx, requestsURL)
+	if err != nil {
+		return nil, err
+	}
+
+	responses, err := pubsub.OpenTopic(ctx, responsesURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Messenger{
+		Deployments: deployments,
+		Endpoints:   endpoints,
+		Queues:      queues,
+		HTTPC:       httpClient,
+		requests:    requests,
+		responses:   responses,
+	}, nil
+}
+
+func (m *Messenger) Start(ctx context.Context) error {
+	for {
+		msg, err := m.requests.Receive(ctx)
+		if err != nil {
+			return err
+		}
+
+		log.Println("Received message:", msg.LoggableID)
+
+		m.handleRequest(context.Background(), msg)
+
+		// Slow down a bit to avoid churning through messages and running
+		// up cloud costs when no meaningful work is being done.
+		if consecutiveErrors := m.getConsecutiveErrors(); consecutiveErrors > 0 {
```

Comment:
I think there is also a risk that occasionally there is a short spike of errors that would slow things down. I think

Reply:
True, we will probably need to tune this over time. I think its a good thing to slow stuff down when errors start building up. Right now the wait time will go back to zero once a single message is processed successfully.

This delay account for spontaneous failures that might creep up overnight, but also is intended to mitigate a failure that might happen if some job send a million malformed requests into a topic.

Reply:
Added comment containing these thoughts